### PR TITLE
HirIdification: rework Map

### PR DESCRIPTION
**Pull request by ljedrz**
The next iteration of HirIdification (#57578).

- remove `NodeId` from `Entry`
- change `Map::map` to an `FxHashMap<HirId, Entry>`
- base the `NodeId` `Map` methods on `HirId` ones (reverses the current state)
- HirIdify `librustdoc` a little bit (some `NodeId` `Map` methods were converted to work on `HirId`s)

The second change might have performance implications, so I'd do a perf run to be sure it's fine; it simplifies the codebase and shouldn't have an impact as long as the `Map` searches are cached (which is now possible thanks to using `HirId`s).

r? @Zoxc

### CODE REVIEW

**Comment by ljedrz** [comment]
Oof

@bors try (for perf, for real this time)

**Comment by ljedrz** [comment]
Can I get a perf run pls?

**Comment by Xanewok** [comment]
Assuming I remember the invocation...

@rust-timer build 3c13d8d736155feccc096cb9709f2e6b4b9dab7f

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 3c13d8d736155feccc096cb9709f2e6b4b9dab7f

**Comment by Xanewok** [comment]
@Mark-Simulacrum huh so only bare command works in the comment body?

**Comment by Mark-Simulacrum** [comment]
No, it should've responded to you somehow. I need to investigate why it didn't.

**Comment by ljedrz** [comment]
Hmm, looks like there are some instructions and max-rss losses.

@Zoxc do you think we should also give it a shot together with https://github.com/rust-lang/rust/pull/57173 or the `map` should remain a `Vec` (I'm not sure how dense `HirId`s would be, perhaps the results would be similar).

**Comment by Zoxc** [comment]
I would wait until https://github.com/rust-lang/rust/pull/58623 lands before trying to replace the existing node map. Until then I'd just focus on getting rid of `NodeId` uses outside the HIR map.

**Comment by ljedrz** [comment]
Ok; can you mark this PR as blocked so it doesn't pop up during PR queue cleanups?

**Comment by Amanieu** [comment]
Is it really necessary to wait for #58623 to land? That PR only touches libstd and doesn't really interact with this PR. The HashMap API is completely unchanged, just the implementation.

**Comment by ljedrz** [comment]
It's not necessary, but the perf penalty of this PR should be more palatable with the new implementation in place (especially max-rss).

**Comment by ljedrz** [comment]
Rebased.

**Comment by Zoxc** [comment]
@ljedrz You can move the rustdoc changes over to https://github.com/rust-lang/rust/pull/59068

**Comment by ljedrz** [comment]
Rebased.

@Zoxc the last few days have been quite busy for me (just as the following few will be) and I didn't manage to do it (it's already rolled up). I should soon be back to my snappy response time, though 👌.

**Comment by Zoxc** [comment]
@ljedrz hashbrown in libstd has landed now, so we can see if it helps performance here.

**Comment by ljedrz** [comment]
Sounds good.

@bors try

**Comment by ljedrz** [comment]
Ready for a perf run.

**Comment by Xanewok** [comment]
@rust-timer build 7378b7bfe6cfb37a5aeadf1054d823473f7f7ddd

File: src/librustc/hir/map/collector.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ impl<'a, 'hir> NodeCollector<'a, 'hir> {
         let mut collector = NodeCollector {
             krate,
             source_map: sess.source_map(),
-            map: repeat(None).take(sess.current_node_id_count()).collect(),
+            map: Default::default(),
```
Comment (Zoxc) [inline src/librustc/hir/map/collector.rs]: You can use the node count as the capacity here still.

Comment (ljedrz) [reply src/librustc/hir/map/collector.rs]: Actually that function is no longer available with this commit; should I bring it back or "inline" it?

Comment (Zoxc) [reply src/librustc/hir/map/collector.rs]: Bring it back. It was just there for this code.

**Comment by ljedrz** [comment]
@Zoxc comment addressed.

**Comment by Zoxc** [comment]
hashbrown didn't seem to help. I guess we'll have to do some optimizations here later.

@bors r+

**Comment by ljedrz** [comment]
I mean the instruction results did seem to improve; or did you expect greens in the incremental department?

**Comment by Zoxc** [comment]
@ljedrz wall-time is the metric which counts, which seems pretty much as bad as before.

**Comment by pietroalbini** [comment]
Yielding priority to the stable release.

@bors retry

**Comment by pietroalbini** [comment]
Yielding priority to the stable release, again.

@bors retry

**Pull request merged**